### PR TITLE
mon: validate implicit erasure code crush ruleset

### PR DESCRIPTION
**Pull request by dachary**
http://tracker.ceph.com/issues/11814

### CODE REVIEW

**Comment by dachary** [comment]
@tchaikov this is on top of your excellent https://github.com/ceph/ceph/pull/4726 :-)

File: src/tools/crushtool.cc [hunk 1]
```diff
@@ -226,6 +227,8 @@ int main(int argc, const char **argv)
   std::string infn, srcfn, outfn, add_name, remove_name, reweight_name;
   bool compile = false;
   bool decompile = false;
+  bool check_names = false;
```
Comment (dachary) [inline src/tools/crushtool.cc]: s/check_names/check/

File: src/tools/crushtool.cc [hunk 2]
```diff
@@ -505,7 +510,7 @@ int main(int argc, const char **argv)
     cerr << "cannot specify more than one of compile, decompile, and build" << std::endl;
     exit(EXIT_FAILURE);
   }
-  if (!compile && !decompile && !build && !test && !reweight && !adjust && !tree &&
+  if (!check_names && !compile && !decompile && !build && !test && !reweight && !adjust && !tree &&
```
Comment (dachary) [inline src/tools/crushtool.cc]: also if (!check line 505 above otherwise it will display a spurious warning

File: src/tools/crushtool.cc (hunk 1 above)
Comment (tchaikov) [inline src/tools/crushtool.cc]: thanks, will do.

File: src/tools/crushtool.cc (hunk 2 above)
Comment (tchaikov) [inline src/tools/crushtool.cc]: good catch. =) will update it accordingly.

**Comment by dachary** [comment]
@t-miyamae do you approve this change  9bac353 ?

**Comment by dachary** [comment]
it needs a test to verify a bogus failure-domain is going to be rejected as expected

**Comment by t-miyamae** [comment]
@dachary I almost understand your intent, but may I confirm just one thing? Should we create at least one ruleset for each different size (=k+m) of pool?

**Comment by dachary** [comment]
@t-miyamae it's not necessary to create a ruleset for each different size (=k+m) of a pool. It is only in the case where the ruleset is created implicitly because the caller did not specify another ruleset to use. In reality if people are using various combinations of k+m for the same plugin, it is likely they will also have a carefully crafted ruleset to match their need and be used for all pools. But I'm speculating really ;-)

**Comment by t-miyamae** [comment]
@dachary I see. Thank you for your explanation. We approve 9bac353.

**Comment by dachary** [comment]
@t-miyamae thanks !

**Comment by dachary** [comment]
rebased and repushed

**Comment by tchaikov** [comment]
@dachary is it possible to add a test for 7109919? where we can repeat Herve's steps in http://tracker.ceph.com/issues/11814 ?

**Comment by dachary** [comment]
@tchaikov ok.

**Comment by tchaikov** [comment]
sorry @ldachary , i withdraw my query about adding a test. i was trying to repeat the steps detailed in http://tracker.ceph.com/issues/11814, but seems the hard part is to inject the ruleset of "castor-ec-isa" into monitor before creating the ecpool.
i tried to compile the attached crush map, and failed:
```
$ ./crushtool -c crush.txt -o crush.11814
in rule 'bigbang' item 'bucket10' not defined
in rule 'eos-ec-jerasure' item 'bucket10' not defined
in rule 'eos-ec-isa' item 'bucket10' not defined
in rule 'testbbecj' item 'bucket10' not defined
in rule 'testbbeci' item 'bucket10' not defined
in rule 'castor-ec-isa' item 'bucket10' not defined
in rule 'eos-cold' item 'bucket10' not defined
in rule 'testbbec-isa' item 'bucket10' not defined
in rule 'testbbec-jerasure' item 'bucket10' not defined
```

could you remove the DNM? and i think it's good to be merged.

**Comment by dachary** [comment]
@tchaikov I may have found a way to reproduce the problem actually. Trying now.

**Comment by dachary** [comment]
@tchaikov in addition I think the test, if needed at all, is not where it should. The real danger is to assign a pool to a bugous ruleset, therefore the test should be done just before doing that, regardless of how the ruleset was created.

**Comment by dachary** [comment]
@tchaikov here is a revised patch. I believe it covers http://tracker.ceph.com/issues/11814 as well as a broader range of problems, should a bugous ruleset reach the point where it can be **used with osd pool create**. And with a test: I'm grateful you reminded me of the test :-)

**Pull request merged**